### PR TITLE
Fix regex backtracking and cross-statement scoping in NoDeleteWithoutWhere and NoGrantExceptGrantExecute

### PR DESCRIPTION
**Pull request by recampbell**
## Summary

- Replaces `(?is)(?=.*\b(keyword)\b)(?!.*\b(companion)\b).*` regex patterns with anchored, per-statement alternatives that fix both a **performance bug** and a **correctness bug**
- Affects `Regex/SQL Server/NoDeleteWithoutWhere.md` and `Regex/SQL Server/NoGrantExceptGrantExecute.md`

### 1. Catastrophic backtracking (performance)

The `(?=.*\b(delete)\b)` positive lookahead forces Java's regex engine to attempt word-boundary matching at every character position across the entire changeset content. On large changesets with thousands of rows (common in bulk DML for insurance/financial workloads), this causes exponential runtime — we measured **7+ hours** on a customer's changelog, and the JVM crashed with SIGBUS at 10K rows per file in our reproduction.

### 2. Cross-statement scope contamination (correctness)

The `(?s)` (DOTALL) flag makes `.*` span the entire changeset, not individual SQL statements. This causes false negatives in multi-statement changesets:

```sql
--changeset example:multi-statement
DELETE FROM OldRecords;                          -- no WHERE — should be flagged
SELECT * FROM AuditLog WHERE action = 'cleanup'; -- WHERE here satisfies the lookahead
```

The `WHERE` in the second statement satisfies the `(?!.*\b(where)\b)` negative lookahead, so the bare `DELETE` passes unflagged. Same issue applies to GRANT/EXECUTE.

## Fix

| Check | Old pattern | New pattern |
|-------|-------------|-------------|
| NoDeleteWithoutWhere | `(?is)(?=.*\b(delete)\b)(?!.*\b(restrict)\b)(?!.*\bno\s+action\b)(?!.*\bset\s+null\b)(?!.*\b(where)\b).*` | `(?i)\bdelete\b(?![^;]*(?:\bwhere\b\|\brestrict\b\|\bno\s+action\b\|\bset\s+null\b))` |
| NoGrantExceptGrantExecute | `(?is)(?=.*\b(grant)\b)(?!.*\b(execute)\b).*` | `(?i)\bgrant\b(?![^;]*\bexecute\b)` |

The new patterns:
- **Anchor** directly on the keyword (`\bdelete\b` / `\bgrant\b`) — no backtracking
- **Scope** the negative lookahead to the current SQL statement via `[^;]*` (up to the next semicolon)
- **Preserve** all existing FK-exclusion logic (RESTRICT, NO ACTION, SET NULL) for the DELETE check

## Verification

Tested against 10 test cases on Liquibase Secure 5.1.0:
- 4 positive tests (should flag) — all pass, including multi-statement changesets
- 4 negative tests (should not flag) — all pass, including FK constraints and string literals
- 2 edge cases (multi-line SQL, mixed case) — all pass

Performance at 10K rows per large file: regex processing dropped from >5 min (JVM crash) to seconds.

## Test plan

- [ ] Verify existing sample passing/failing scripts in both `.md` files still produce the documented behavior
- [ ] Test multi-statement changesets (new sample scripts added to both files)
- [ ] Confirm no regression on FK constraint exclusions (RESTRICT, NO ACTION, SET NULL)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: Regex/SQL Server/NoDeleteWithoutWhere.md [hunk 1]
```diff
@@ -21,11 +39,23 @@ ALTER TABLE address ADD CONSTRAINT fk_address_city FOREIGN KEY (city_id) REFEREN
 --changeset mikeo:delete-setnull
 ALTER TABLE payment ADD CONSTRAINT payment_rental_id_fkey FOREIGN KEY (rental_id) REFERENCES rental (rental_id) ON UPDATE CASCADE ON DELETE SET NULL
 ```
+``` sql
+--changeset mikeo:delete-multi-statement
+-- DELETE without WHERE should still be caught even when WHERE appears in another statement
+DELETE FROM dbo.Table01;
+SELECT * FROM dbo.AuditLog WHERE Action = 'cleanup';
+```
```
[the remark this answers is not in the document]
Comment (recampbell) [reply Regex/SQL Server/NoDeleteWithoutWhere.md]: Agree, updated

File: Regex/SQL Server/NoDeleteWithoutWhere.md [hunk 2]
```diff
@@ -47,6 +71,6 @@ Message:            Error! All DELETE statements must have a WHERE clause.
 | > | `liquibase checks customize --check-name=SqlUserDefinedPatternCheck` |
 | Give your check a short name for easier identification (up to 64 alpha-numeric characters only) [SqlUserDefinedPatternCheck1]: | `NoDeleteWithoutWhere` |
 | Set the Severity to return a code of 0-4 when triggered. (options: 'INFO'=0, 'MINOR'=1, 'MAJOR'=2, 'CRITICAL'=3, 'BLOCKER'=4)? [INFO]: | `<Choose a value: 0, 1, 2, 3, 4>` |
-| Set 'SEARCH_STRING' (options: a string, or a valid regular expression): | `(?is)(?=.*\b(delete)\b)(?!.*\b(restrict)\b)(?!.*\bno\s+action\b)(?!.*\bset\s+null\b)(?!.*\b(where)\b).*` |
+| Set 'SEARCH_STRING' (options: a string, or a valid regular expression): | `(?i)\bdelete\b(?![^;]*(?:\bwhere\b|\brestrict\b|\bno\s+action\b|\bset\s+null\b))` |
```
Comment (daticalamy) [inline Regex/SQL Server/NoDeleteWithoutWhere.md:74]: Hi @recampbell : the SEARCH_STRING is within a table so the pipes have to be escaped. If you look here https://github.com/recampbell/custom_policychecks/blob/fix/regex-backtracking-and-scoping/Regex/SQL%20Server/NoDeleteWithoutWhere.md it doesn't display correctly.
